Libraries/dotNetRDF/Writing/GraphMLWriter.cs

```diff
@@ -0,0 +1,282 @@
+            }
+
+            // RDF is always a directed graph
+            writer.WriteAttributeString(GraphMLHelper.Edgedefault, GraphMLHelper.Directed);
+
+            // The next two iterations are disjoint
+            foreach (var literalNode in graph.Nodes.Where(n => n.NodeType != NodeType.Literal))
+            {
+                GraphMLWriter.WriteNode(writer, literalNode.GetHashCode().ToString(), literalNode.ToString());
+            }
+
+            // Literal node identifiers are the whole statement so they become distinct
+            foreach (var triple in graph.Triples.Where(t => t.Object.NodeType == NodeType.Literal))
+            {
+                GraphMLWriter.WriteNode(writer, triple.GetHashCode().ToString(), triple.Object.ToString());
+            }
+
+            foreach (var triple in graph.Triples)
+            {
+                GraphMLWriter.WriteEdge(writer, triple);
+            }
+
+            writer.WriteEndElement();
+        }
+
+        private static void WriteEdge(XmlWriter writer, Triple triple)
+        {
+            writer.WriteStartElement(GraphMLHelper.Edge);
+            writer.WriteAttributeString(GraphMLHelper.Source, triple.Subject.GetHashCode().ToString());
+
+            writer.WriteStartAttribute(GraphMLHelper.Target);
+            if (triple.Object.NodeType == NodeType.Literal)
+            {
+                writer.WriteString(triple.GetHashCode().ToString());
+            }
+            else
+            {
+                writer.WriteString(triple.Object.GetHashCode().ToString());
+            }
+
+            GraphMLWriter.WriteData(writer, GraphMLHelper.EdgeLabel, triple.Predicate.ToString());
+
+            writer.WriteEndElement();
+        }
+
+        private static void WriteKey(XmlWriter writer, string id, string domain)
+        {
+            writer.WriteStartElement(GraphMLHelper.Key);
+            writer.WriteAttributeString(GraphMLHelper.Id, id);
+            writer.WriteAttributeString(GraphMLHelper.Domain, domain);
+            writer.WriteAttributeString(GraphMLHelper.AttributeTyte, GraphMLHelper.String);
+            writer.WriteEndElement();
+        }
+
+        private static void WriteNode(XmlWriter writer, string id, string label)
+        {
+            writer.WriteStartElement(GraphMLHelper.Node);
+            writer.WriteAttributeString(GraphMLHelper.Id, id);
+
+            GraphMLWriter.WriteData(writer, GraphMLHelper.NodeLabel, label);
+
+            writer.WriteEndElement();
+        }
+
+        private static void WriteData(XmlWriter writer, string key, string value)
+        {
+            writer.WriteStartElement(GraphMLHelper.Data);
+            writer.WriteAttributeString(GraphMLHelper.Key, key);
+            writer.WriteString(value);
+            writer.WriteEndElement();
+        }
+    }
+
+    /// <summary>
+    /// A helper class containing GraphML name and URI constants
+    /// </summary>
+    public static class GraphMLHelper
```

Comment:
Please put classes in individual files

Reply:
Besides, I don't particularly like the name but I think we have other similar classes for Xsd, etc, right? @kal maybe we could put `GraphMLHelper` in the same namespace?

Reply:
@tpluscode, @langsamu 

+1 for separate files for each class. I know that this is not the case with the rest of the code-base and I think it is probably a matter of personal style, but I find it difficult to work with code where multiple classes are combined in a single source file.

The xxx**SpecsHelper** name is quite a common pattern in the code base, so can we please rename this class to `GraphMLSpecsHelper`? Many of those `xxxSpecsHelper` classes sit in the Parsing namespace as they are used by both parsers and writers. In this case we won't have a GraphML parser, so I'm OK with the GraphMLSpecsHelper going in the Writing namespace.

Reply:
Fair enough, let's keep it here.